internal/secrets/secrets.go

```
@@ -0,0 +1,165 @@
+		if !secretsManagerAPI.IsEnabled() {
+			secretsManagerAPI.Setup()
+		}
+
+		r := strings.ReplaceAll(secret, "SECRETS_AWS:", "")
+		v, err := secretsManagerAPI.GetSecret(ctx, r)
+		if err != nil {
+			return "", fmt.Errorf("secrets %s: %v", secret, err)
+		}
+
+		if err := cache.SetWithTTL(ctx, secret, v, ttl); err != nil {
+			return "", fmt.Errorf("secrets %s: %v", secret, err)
+		}
+
+		return v, nil
+	}
+
+	return "", fmt.Errorf("secrets %s: %v", secret, errSecretNotFound)
+}
+
+// Interpolate identifies when a string contains one or more secrets and
+// interpolates each secret with the string. This function uses the same
+// convention as the standard library's regexp package for capturing named
+// groups (${name}).
+//
+// For example, if the string is "/path/to/{SECRETS_ENV:FOO}" and BAR is the
+// secret stored in the environment variable FOO, then the interpolated string
+// is "/path/to/BAR".
+//
+// Multiple secrets can be stored in a single string; if the string is
+// "/path/to/{SECRETS_ENV:FOO}/{SECRETS_ENV:BAZ}", then the interpolated string
+// is "/path/to/BAR/QUX".
+//
+// If more than one interpolation function is applied to a string (e.g., non-secrets
+// capture groups), then this function must be called first.
+func Interpolate(ctx context.Context, s string) (string, error) {
+	if !strings.Contains(s, "${SECRETS_") {
+		return s, nil
+	}
+
+	re, err := regexp.Compile(interpRe)
+	if err != nil {
+		return "", err
+	}
+
+	matches := re.FindAllStringSubmatch(s, -1)
+	for _, m := range matches {
+		if len(m) == 0 {
+			continue
+		}
+
+		secret, err := Get(ctx, m[len(m)-1])
+		if err != nil {
+			return "", err
+		}
+
+		// replaces each substring with a secret.
+		// if the secret is BAR and the string was
+		// "/path/to/secret/{SECRETS_ENV:FOO}", then the interpolated
+		// string output is "/path/to/secret/BAR".
+		old := fmt.Sprintf("${%s}", m[len(m)-1])
+		s = strings.Replace(s, old, secret, 1)
+	}
+
+	return s, nil
+}
+
+func init() {
```

Should the default behavior when a cache fails to initialize be operating without a cache? (Though tbh I don't think it's possible to fail the Setup for the memory KV anyways)

---

it's probably impossible to fail, but i think it's better to panic so the error is surfaced instead of silently slowing down data processing (in case a remote secrets backend is used)